### PR TITLE
Fixes the build issues on pkg-build.racket-lang.org for AlexKnauth/afl

### PR DESCRIPTION
**Pull request by SuzanneSoy**
Fixes the build issues on pkg-build.racket-lang.org for AlexKnauth/afl, which were happening because builds are not run in isolation when loading reader modules.

Please review this patch with care, as I am not 100% sure it won't introduce any regressions.

See https://github.com/AlexKnauth/afl/issues/1#issuecomment-257298961 for an explanation of why I think this fix is correct.

### CODE REVIEW

**Comment by AlexKnauth** [comment]
What if I want the `read-syntax` function to be able to depend in some way on the old `get-info` function? Is there a way to do that?

**Comment by SuzanneSoy** [comment]
With the patch it is not possible, but it was not really possible earlier either: as the sequence of `?set!` statements showed, only `read-syntax` was set during compilation.

With the former implementation, if I understood properly, it would be possible to do it in a somewhat convoluted way:
- First get the result of `lang-extension->lang-reader` (which is the result of a call to [`make-meta-reader`](http://docs.racket-lang.org/syntax/reader-helpers.html#%28def._%28%28lib._syntax%2Fmodule-reader..rkt%29._make-meta-reader%29%29), wrapped in a struct)
- Then call `get-info` on that with the appropriate input stream, causing make-meta-reader to consume the "wrapped language name" (e.g. `the-wrapped-language` in `#lang afl the-wrapped-language more-stuff…`), and causing the `?get-info` delayed variable to be set
- Then call `read-syntax`, which will be able to use the freshly set `?get-info`.

In other words, in order for `read-syntax` to be able to rely on `get-info`, it is necessary to use `get-info` before calling `read-syntax`, and rely on the side-effects of that operation. That seems like a rather unreliable usage pattern (and incidentally, a file using such a language would likely not compile at all, because the compiler does not seem to call `get-info`.

In order to be able to have `read-syntax` depend on `get-info` in a reliable fashion, I think it would be necessary to partially rewrite `make-meta-reader`, to pass all three "old" functions to each of the `convert-read`, `convert-read-syntax` and `convert-get-info` functions.

**Comment by AlexKnauth** [comment]
There are three "conversion" functions that I pass to `make-meta-reader`, which are `convert-read`, `convert-read-syntax`, and `convert-get-info`. When `make-meta-reader` receives those functions, it needs to apply all three of those functions to their respective `read`, `read-syntax` and `get-info` functions from the base language. When it calls `(convert-read-syntax read-syntax)`, I need to return a function, and I can do that. Then when it calls `(convert-get-info get-info)`, I can save the `get-info` function so that the function I returned already from `convert-read-syntax` can use it. 

That isn't ideal; the right thing to do would be to rewrite `make-meta-reader` entirely. This summer I started trying to do that, but I got confused very quickly, and soon I was too busy working on more important things.

**Comment by AlexKnauth** [comment]
> because the compiler does not seem to call `get-info`

Oh, I see what you're saying now. The `make-meta-reader` call doesn't apply those functions right away; it returns functions that call them when needed. That explains some of the weird behavior I've been getting, but not this particular issue. 

However, that is enough to show that there really is no way to do that without rewriting `make-meta-reader`, so this patch doesn't actually lose anything.
